<a href="https://colab.research.google.com/github/18521157/CS114.K21/blob/master/round3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Lấy dữ liệu:**
Dùng hàm loadtxt từ thư viện numpy để load file từ http://cs.uit.edu.vn/data3.txt, sau đó in số dòng, số cột của dataset.

In [0]:
from numpy import loadtxt
from urllib.request import urlopen
url = 'http://cs.uit.edu.vn/data3.txt'
raw_data = urlopen(url)
dataset = loadtxt(raw_data, delimiter=",")
print(dataset.shape)

(10590, 2)


**Chia dataset :** Dùng hàm train_test_split từ thư viện sklearn để chia dataset thành 2 phần: 80% train và 20% test.

In [0]:
from sklearn.model_selection import train_test_split
X, y = dataset[:, :1], dataset[:, -1]
print(X)
print(y)
X_train, X_test, Y_train, Y_test = train_test_split(X, y, test_size=0.20, random_state=1)

[[37837612.]
 [37395246.]
 [36610827.]
 ...
 [33520323.]
 [21243264.]
 [21380484.]]
[4.12166 3.92089 3.2154  ... 2.90918 1.80292 1.82795]


**Chọn model và train:** Chọn model LinearRegresion từ sklearn để train.



In [0]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

**Dự đoán:**

In [0]:
pred = model.predict(X_test)

In [0]:
import sklearn.metrics as metrics
print('MSE:',metrics.mean_squared_error(Y_test,pred))
print('train:', model.score(X_train,Y_train))
print('test:', model.score(X_test,Y_test))


MSE: 0.0029732305247697818
train: 0.9877941176410101
test: 0.9897290008911028


**Ta thấy:** Vấn đề tệ hơn: MSE(data3) = 0,00297 => lớn hơn MSE(data2)) = 0.000144.
**Giải thích:**

# **Xây dựng model mới và ứng dụng dự báo mới sử dụng phương pháp tăng số features đã gợi ý trên lớp:**

**Lấy dữ liệu:** Load dataset từ http://cs.uit.edu.vn/data3.txt bằng hàm loadtxt từ numpy.
Sau đó đổi type của dataset từ numpy.ndarray => pandas.core.frame.DataFrame.

In [1]:
from numpy import loadtxt
import pandas as pd
from urllib.request import urlopen
url = 'http://cs.uit.edu.vn/data3.txt'
raw_data = urlopen(url)
dataset = loadtxt(raw_data, delimiter=",")
print(dataset.shape)
newdataset=pd.DataFrame(data=dataset)
newdataset.columns = ["size","time"]

(10590, 2)


**Thêm 2 features : size^2 và log(size)**

In [2]:
import math
df=pd.DataFrame(newdataset)
Size2=df['size']**2
Sizelog=[df['size'][i]*math.log(df['size'][i]) for i in range(len(newdataset))]
df['Size2']=Size2
df['Sizelog']=Sizelog
print(newdataset)

             size     time         Size2       Sizelog
0      37837612.0  4.12166  1.431685e+15  6.602215e+08
1      37395246.0  3.92089  1.398404e+15  6.520629e+08
2      36610827.0  3.21540  1.340353e+15  6.376088e+08
3      29555651.0  2.54506  8.735365e+14  5.084100e+08
4      25928216.0  2.21831  6.722724e+14  4.426165e+08
...           ...      ...           ...           ...
10585  26872494.0  2.35117  7.221309e+14  4.596974e+08
10586  20753811.0  1.79155  4.307207e+14  3.496652e+08
10587  33520323.0  2.90918  1.123612e+15  5.808288e+08
10588  21243264.0  1.80292  4.512763e+14  3.584068e+08
10589  21380484.0  1.82795  4.571251e+14  3.608596e+08

[10590 rows x 4 columns]


**Chia dataset :** Dùng hàm train_test_split từ thư viện sklearn để chia dataset thành 2 phần: 80% train và 20% test.

In [0]:
from sklearn.model_selection import train_test_split
X2=newdataset[["size","Size2","Sizelog"]]
y2=newdataset["time"].values.reshape(-1,1)
X_train2, X_test2, Y_train2, Y_test2 = train_test_split(X2, y2, test_size=0.20, random_state=1)

**Chọn model và train:** Chọn model LinearRegresion từ sklearn để train.

In [4]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train2, Y_train2)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

**Dự đoán:**

In [0]:
pred2 = model.predict(X_test2)

In [6]:
import sklearn.metrics as metrics
print('MSE2:',metrics.mean_squared_error(Y_test2,pred2))
print('train2:', model.score(X_train2,Y_train2))
print('test2:', model.score(X_test2,Y_test2))

MSE2: 0.0029729965832469333
train2: 0.9878348218741277
test2: 0.9897298090400682


**So với các kết quả cũ:**

*   MSE(cũ) : 0.0029732305247697818
*   MSE(mới): 0.0029729965832469333 (giảm)

*   train(cũ):  0.9877941176410101
*   train(mới): 0.9878348218741277 (tăng)

*   test(cũ):  0.9897290008911028
*   test(mới): 0.9897298090400682 (tăng)




